openssl/src/ssl/mod.rs

```diff
@@ -3704,6 +3709,49 @@ impl<S: Read + Write> SslStream<S> {
         }
     }
 
+    /// Like `peek`, but returns an `ssl::Error` rather than an `io::Error`.
+    ///
+    /// It is particularly useful with a nonblocking socket, where the error value will identify if
+    /// OpenSSL is waiting on read or write readiness.
+    ///
+    /// This corresponds to [`SSL_peek`].
+    ///
+    /// [`SSL_peek`]: https://www.openssl.org/docs/manmaster/man3/SSL_peek.html
+    pub fn ssl_peek(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        // See above for why we short-circuit on zero-length buffers
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        let ret = self.ssl.peek(buf);
+        if ret > 0 {
+            Ok(ret as usize)
+        } else {
+            Err(self.make_error(ret))
+        }
+    }
+
+    /// Receives data on the socket from the remote address to which it is connected, without removing that data from the queue.
+    ///
+    /// Successive calls return the same data.
+    pub fn peek(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
I don't think this variant is necessary since we don't need to conform to an existing trait.

Reply:
Yeah, you're right. First I thought, it'd be rather convenient to have a method which behaves like the corresponding method of `TcpStream` (as `.read()` and `.write()` do). But if no trait requires it, it is unnecessary to "ignore" possible error codes...

I'll remove it and update the PR sometime later today :)